File: Javascript/string_to_integer.js
/**
 * @param {string} s
 * @return {number}
 */
function isdigit(char) {
  if (typeof char !== 'string') {
    return false;
  }

  if (char.trim() === '') {
    return false;
  }

  return !isNaN(char);
}

var myAtoi = function(s) {
    var num = 0
    var check_sign = false
    var check_digit = false
    var sign = 1
    var MAX_INT = Math.pow(2, 31) - 1
    var MIN_INT = -MAX_INT - 1
    for (i in s) {
        if (s[i] === ' ' && !check_sign && !check_digit) continue
        if (!isdigit(s[i])) {
            console.log(s[i])
            if (['-', '+'].includes(s[i]) === false || check_sign || check_digit) {
                break
            }
        }
        else {
            if (!check_digit) check_digit = true
            num = num * 10 + parseInt(s[i])
        }
        
        if (['-', '+'].includes(s[i])) {
            check_sign = true
            if (s[i] === '-') {
                sign = -1
            }
        }
    }

    num = sign * num
    if (num < MIN_INT) return MIN_INT
    if (num > MAX_INT) return MAX_INT
    return num
};

File: Python/string_to_integer.py
class Solution():
    def __init__(self):
        self.name = 'S->13'

    # Time complexity: O(n)
    # Space complexity: O(1)
    def StringToInteger(self, s):
        sign = 1
        check_digit = False
        check_sign = False
        num = 0
        INT_MAX = pow(2, 31) - 1
        INT_MIN = -pow(2, 31)

        for m in s:
            if m == ' ' and not check_digit and not check_sign:
                continue
            if not m.isdigit():
                if m not in ['-', '+'] or check_sign or check_digit:
                    break
            else:
                check_digit = True
                num = num * 10 + int(m)
            if m in ['-', '+']:
                if m == '-':
                    sign = -1
                check_sign = True
        num = sign * num
        if num < INT_MIN:
            return INT_MIN
        if num > INT_MAX:
            return INT_MAX

        return num

print(Solution().StringToInteger(' + 123'))
